components/script/dom/bindings/buffer_source.rs

```diff
@@ -226,27 +216,138 @@ where
                 let mut is_shared = false;
                 unsafe {
                     if JS_IsArrayBufferViewObject(*buffer.handle()) {
-                        // If it is an ArrayBuffer view, get the buffer using JS_GetArrayBufferViewBuffer
                         rooted!(in (*cx) let view_buffer =
                             JS_GetArrayBufferViewBuffer(*cx, buffer.handle(), &mut is_shared));
-                        // This buffer is always created unshared
                         debug_assert!(!is_shared);
-                        // Detach the ArrayBuffer
-                        DetachArrayBuffer(*cx, view_buffer.handle())
+                        return GetArrayBufferByteLength(*view_buffer.handle());
                     } else {
-                        // If it's not an ArrayBuffer view, Detach the buffer directly
-                        DetachArrayBuffer(*cx, Handle::from_raw(buffer.handle()))
+                        return GetArrayBufferByteLength(*buffer.handle());
                     }
                 }
             },
+            _ => unreachable!(
+                "viewed_buffer_array_byte_length should only be called on ArrayBufferView"
+            ),
         }
     }
 
+    pub fn byte_length(&self) -> usize {
+        match &self.buffer_source {
+            BufferSource::Int8Array(buffer) |
+            BufferSource::Int16Array(buffer) |
+            BufferSource::Int32Array(buffer) |
+            BufferSource::Uint8Array(buffer) |
+            BufferSource::Uint16Array(buffer) |
+            BufferSource::Uint32Array(buffer) |
+            BufferSource::Uint8ClampedArray(buffer) |
+            BufferSource::BigInt64Array(buffer) |
+            BufferSource::BigUint64Array(buffer) |
+            BufferSource::Float32Array(buffer) |
+            BufferSource::Float64Array(buffer) |
+            BufferSource::Default(buffer) |
+            BufferSource::DataView(buffer) => unsafe {
+                JS_GetArrayBufferViewByteLength(*buffer.handle())
+            },
+            BufferSource::ArrayBuffer(buffer) => unsafe {
+                GetArrayBufferByteLength(*buffer.handle())
+            },
+        }
+    }
+
+    pub fn array_length(&self) -> usize {
+        self.get_buffer().unwrap().len()
+    }
+
+    pub(crate) fn has_typed_array_name(&self) -> bool {
+        match &self.buffer_source {
+            BufferSource::Int8Array(_) |
+            BufferSource::Int16Array(_) |
+            BufferSource::Int32Array(_) |
+            BufferSource::Uint8Array(_) |
+            BufferSource::Uint16Array(_) |
+            BufferSource::Uint32Array(_) |
+            BufferSource::Uint8ClampedArray(_) |
+            BufferSource::BigInt64Array(_) |
+            BufferSource::BigUint64Array(_) |
+            BufferSource::Float32Array(_) |
+            BufferSource::Float64Array(_) |
+            BufferSource::ArrayBuffer(_) |
+            BufferSource::Default(_) => true,
```

Comment:
I think this is incorrect, because the buffer is not a typed array, see https://tc39.es/ecma262/#typedarray

So basically, if it is a view, and it is not a data view, then true.